motionplan/rrtStarConnect.go

```diff
@@ -165,45 +150,56 @@ func (mp *rrtStarConnectMotionPlanner) rrtBackgroundRunner(ctx context.Context,
 
 		// try to connect the target to map 1
 		utils.PanicCapturingGo(func() {
-			mp.extend(algOpts, rrt.rm.startMap, target, m1chan)
+			mp.extend(rrt.rm.startMap, target, m1chan)
 		})
 		utils.PanicCapturingGo(func() {
-			mp.extend(algOpts, rrt.rm.goalMap, target, m2chan)
+			mp.extend(rrt.rm.goalMap, target, m2chan)
 		})
 		map1reached := <-m1chan
 		map2reached := <-m2chan
 
 		if map1reached != nil && map2reached != nil {
 			// target was added to both map
-			solved = true
 			shared = append(shared, &nodePair{map1reached, map2reached})
+
+			// Check if we can return
+			if nSolved%defaultOptimalityCheckIter == 0 {
+				solution := shortestPath(rrt.rm, shared)
+				solutionCost := EvaluatePlan(solution, mp.algOpts.plannerOptions)
+				if solutionCost-rrt.rm.optNode.cost < defaultOptimalityThreshold*rrt.rm.optNode.cost {
+					mp.logger.Debug("RRT* progress: sufficiently optimal path found, exiting")
+					rrt.solutionChan <- solution
+					return
+				}
+			}
+
+			nSolved++
 		}
 
 		// get next sample, switch map pointers
 		target = referenceframe.RandomFrameInputs(mp.frame, mp.randseed)
 	}
 	mp.logger.Debug("RRT* exceeded max iter")
-	rrt.solutionChan <- shortestPath(rrt.rm, shared, optimalCost)
+	rrt.solutionChan <- shortestPath(rrt.rm, shared)
 }
 
 func (mp *rrtStarConnectMotionPlanner) extend(
-	algOpts *rrtStarConnectOptions,
 	tree rrtMap,
 	target []referenceframe.Input,
 	mchan chan node,
 ) {
-	if validTarget := mp.checkInputs(algOpts.planOpts, target); !validTarget {
+	if validTarget := mp.checkInputs(target); !validTarget {
 		mchan <- nil
 		return
 	}
 	// iterate over the k nearest neighbors and find the minimum cost to connect the target node to the tree
-	neighbors := kNearestNeighbors(algOpts.planOpts, tree, target, algOpts.NeighborhoodSize)
+	neighbors := kNearestNeighbors(mp.algOpts.plannerOptions, tree, target, mp.algOpts.NeighborhoodSize)
```

Comment:
[minor] Now that options are stored in a planner its weird that they are also in the alg opts.  We should remove them from alg opts so that they aren't redundant

Reply:
Good point; done